### ISSUE TITLE
Feature Request: Have get/take return Option<T> to reduce panics

### ISSUE THREAD

**Issue by heartsucker** [stepancheg/rust-protobuf#173]
Accessing nested fields like the following is error prone, and not very ergonomic.

```rust
if foo.has_bar() {
    let bar = foo.get_bar();
    if bar.has_baz() {
        let baz = bar.get_baz();
        println!("{:?}", baz);
    }
}
```

It would be preferable to access fields like:

```rust
foo.get_bar()
    .and_then(|bar| bar.get_baz())
    .map(|baz| println!("{:?}", baz))
```

Edit: This seems to be a bit of a duplicate of #153

**Comment by stepancheg** [comment]
`foo.get_bar().get_baz()` does not panic, but returns type default value (e. g. `0` for integers).

**Comment by heartsucker** [comment]
Sorry for the confuion. I hadn't actually encountered a `panic` because I was writing many nested levels of guards, but I saw the `panic`s in the generated code.

I tested this by sending messages with nested `oneof`s, and the problem I have now is I have messages that can have empty content, and at runtime without checking `has_field`, I could accidentally assumed I have an empty message when in fact someone sent a message of a different type. I would like to second the sentiment in #153 that the `enum` types should be exposed so we can have compile-time checking that we're parsing the right message.
